### ISSUE TITLE
Questions about the effective rank.

### ISSUE THREAD

**Issue by GuanlinLee** [yolky/adversarial_ntk_evolution#1]
Thanks for your very interesting paper. It is a good work. However, I found you use the eigenvalues to calculate the effective rank, instead of the singular values in its paper [1]. I am not sure whether this modification is correct. But the eigenvalues can be all positive if and only if the matrix is a positive definite one. So, the probability and the entropy can be nonsense. Could you check whether it is a typo or you incorrectly implement the effective rank? Thanks.

[1] O. Roy and M. Vetterli. The effective rank: A measure of effective dimensionality.

**Comment by yolky** [comment]
Yes we used the eigenvalues rather than the singular values but in our case they are equal as we are working with kernel matrices which are positive semi-definite. Note that we can write the NxN finite NTK matrix as K = G'G, where G is the PxN matrix corresponding to the gradient embeddings of the dataset (with P = model parameter count, N = number of training points).

**Comment by GuanlinLee** [comment]
Do you mean that you calculate the kernel K(x, x'), where x=x'? I think only in this case the kernel is a positive semi-definite matrix. So, if I calculate the kernel K(x, x'), where x is from the training set and x' is from the test set, making the matrix not a positive semi-definite one, the eigenvalues will be in a complex form, including scaling and rotating.

It seems that all previous works focus on using a semi-definite kernel matrix to make it possible to leverage kernel machine learning theory to analyze the generalization. Have you done anything like this? Would you like to share some insight from your experiments?

**Comment by yolky** [comment]
Yes, I forgot to mention that we set x = x', so that we have a psd matrix. The specifics of how we calculate the kernels used in section 3 of the paper are available in ```calculate_kernel.py```. Specifically, the procedure we use is:

1. Sample 500 _training_ points from the dataset (equally spread between classes)
2. Calculate the finite-width NTK after the given epoch, where x and x' are both the same 500 training points sampled in step 1, resulting in a symmetric 500x500 matrix
3. For multi-class problems, we calculate the _trace kernel_ (i.e. averaged over classes), except for when we calculate the mean-kernel-specialization, in which case we use class-specific kernels

We sample only 500 training points as the kernels are quite expensive to compute. The exact results of section 3 could change depending if we sample test points rather than training points, if we perform adversarial perturbations on these points before computing the kernel, or, like you mentioned, set x != x'. Indeed when x != x', an interesting experiment would be to set x to be benign training points, and x' the same training points with adversarial perturbations, although we did not perform this experiment. I cannot speculate how these changes would effect the results.

In our paper we did not explicitly calculate/derive any generalization bounds given these kernel matrices, although we do empirically evaluate the test error of learning on top of these kernels for benign/adversarial training in section 4.

If you are interested, I may be able to share some of the models/kernel matrices obtained from our experiments, since some of these experiments take some time to run. The file are quite large though. If you are interested in this please contact my email at loo@mit.edu .

**Comment by GuanlinLee** [comment]
Thanks for your explanation. It is clear now. I will try to contact you through email.

**Comment by GuanlinLee** [comment]
I meet another question, when implementing the KSM. In your paper, you say that the y_c is the one hot label. Do you mean that y_c is a matrix (N,K), where the c-th column is 1 and others are zero? But the result of y_c y_cT will be an N*N matrix with 1 in every position. I think I misunderstand what you mean. Could you help me with it?

**Comment by yolky** [comment]
y_c is the one-hot label for class c. So it is a NxK matrix, where in row n, and column k, it is 1 if x_n is in class c, and 0 if x_n is any other class. This makes y_c y_cT, be NxN, with entry i,j == 1 if both x_i and x_j are in class c, and 0 otherwise. See [1] for more details.

[1] Shan, H., & Bordelon, B. (2021). A Theory of Neural Tangent Kernel Alignment and Its Influence on Training. doi:10.48550/ARXIV.2105.14301

**Comment by GuanlinLee** [comment]
Ok, I understand now. The value of the position (n,c) in y_c is depended on the corresponding x_n. If x_n belongs to class c then the value is 1 otherwise 0. So, in the calculation, the c' influences the 1 in the y_c'.

**Comment by GuanlinLee** [comment]
So, if x_n does not belong to class c, the whole row n will be all 0?

**Comment by yolky** [comment]
My apologies I made an error in the previous description. y_c is Nx1, with row n being 1 if x_n is in class c.

Heres an example:

Let's say we have 3 classes and 5 data points. The class labels are [0, 1, 0, 2, 1]. In this case we have:

y_0 = [[1], [0], [1], [0], [0]]
y_1 = [[0], [1], [0], [0], [1]]
y_2 = [[0], [0], [0], [1], [0]]

**Comment by GuanlinLee** [comment]
Thanks a lot. Now, I understand. It is very confusing when the first time reading this chapter.
